src/reaction.jl

```diff
@@ -160,83 +164,74 @@ struct Reaction{S, T}
-        throw(ArgumentError("A reaction requires a non-nothing substrate or product vector."))
-    (isnothing(prodstoich) && isnothing(substoich)) &&
-        throw(ArgumentError("Both substrate and product stochiometry inputs cannot be nothing."))
-
-    if isnothing(subs)
-        prodtype = typeof(value(first(prods)))
-        subs = Vector{prodtype}()
-        !isnothing(substoich) &&
-            throw(ArgumentError("If substrates are nothing, substrate stoichiometries have to be so too."))
-        substoich = typeof(prodstoich)()
-    else
-        subs = value.(subs)
-    end
+function Reaction(rate, subs::Vector, prods::Vector, substoich::Vector{S}, prodstoich::Vector{T};
+                   netstoich = [], metadata = Pair{Symbol, Any}[], 
+                   only_use_rate = metadata_only_use_rate_check(metadata), kwargs...) where {S,T}
+    # Error checks.
+    isempty(subs) && isempty(prods) &&
+        throw(ArgumentError("A reaction requires either a non-empty substrate or product vector."))
+    length(subs) != length(substoich) &&
+        throw(ArgumentError("The substrate vector ($(subs)) and the substrate stoichiometry vector ($(substoich)) must have equal length."))
+    length(prods) != length(prodstoich) &&
+        throw(ArgumentError("The product vector ($(prods)) and the product stoichiometry vector ($(prodstoich)) must have equal length."))
     allunique(subs) ||
         throw(ArgumentError("Substrates can not be repeated in the list provided to `Reaction`, please modify the stoichiometry for any repeated substrates instead."))
-    S = eltype(substoich)
-
-    if isnothing(prods)
-        prods = Vector{eltype(subs)}()
-        !isnothing(prodstoich) &&
-            throw(ArgumentError("If products are nothing, product stoichiometries have to be so too."))
-        prodstoich = typeof(substoich)()
-    else
-        prods = value.(prods)
-    end
     allunique(prods) ||
         throw(ArgumentError("Products can not be repeated in the list provided to `Reaction`, please modify the stoichiometry for any repeated products instead."))
-    T = eltype(prodstoich)
 
-    # try to get a common type for stoichiometry, using Any if have Syms
+    # Ensures everything have uniform and correct types.
+    subs = promote_reaction_vector(subs, BasicSymbolic{Real})
+    prods =  promote_reaction_vector(prods, BasicSymbolic{Real})
     stoich_type = promote_type(S, T)
-    if stoich_type <: Num
-        stoich_type = Any
-        substoich′ = Any[value(s) for s in substoich]
-        prodstoich′ = Any[value(p) for p in prodstoich]
-    else
-        substoich′ = (S == stoich_type) ? substoich : convert.(stoich_type, substoich)
-        prodstoich′ = (T == stoich_type) ? prodstoich : convert.(stoich_type, prodstoich)
-    end
+    (stoich_type <: Num) && (stoich_type = Any)
+    substoich = promote_reaction_vector(substoich, stoich_type)
+    prodstoich = promote_reaction_vector(prodstoich, stoich_type)
```

Comment:
This doesn't seem functionally equivalent to the old code? Why this change? (It is specifically missing the `convert` calls?)